design/open/gc-plus-execution-plan.md

```diff
@@ -0,0 +1,78 @@
+# Uncommitted Garbage Collection - Execution Plan
+
+Uncommitted Garbage Collection [Proposal](link-to-proposal)
+
+## Overview
+
+Implementing uncommitted GC can be divided into 3 main changes:
+1. lakeFS additions and modifications
+2. Implementing uncommitted GC logic on client side
+3. Modifying existing GC
+
+All changes in lakeFS are independent changes, work on lakeFS changes can start immediately.
+Some GC changes are dependent on lakeFS changes, these are prioritized as part of the plan.  
+All tasks are tracked under the [GC+ label](https://github.com/treeverse/lakeFS/labels/GC%2B) in GitHub
+
+## Changes on lakeFS
+
+The following are a list of changes in lakeFS by order of priority, according the dependency constraints 
+
+1. Implement PrepareUncommittedForGC API
+   - Creates metaranges and ranges for a selected branch's staging area, and writes them to a designated GC path
+   - After this is done, we can implement new uncommitted logic in GC
+
+2. Modify [Get/Link]PhysicalAddress
+    - Issue validation token on GetPhysicalAddress with an expiry time and add it to valid token list
+    - LinkPhysicalAddress will validate token and remove it from valid token list
+
+3. Implement Move/RenameObject API
+    - Get entry from branch
+    - Stage new entry using old entry data and new path 
+    - Perform delete for old entry (write tombstone)
+
+4. Modify Gateway CopyObject API
+    - For copy operations on different branches, change CopyObject behavior to perform "full copy" using the underlying 
+   storage adapter's Copy method.
+
+5. Modify lakeFSFS renameObject method
+    - Use CopyObject + DeleteObject when working via Gateway
+    - Use RenameObject when working with OpenAPI
+
+6. lakeFS clients
+    - Any action required on clients?
+    - Do we need to make them version aware?
+    - Should we issue a notification to upgrade clients?
+
+7. Modify StageObject API
+    - Return error if given address is inside repository namespace
+
+8. Implement new repository structure
+    - Root prefix for new lakeFS data - `data/`
+    - Use slice naming conventions as defined in proposal to create object path
+    - Track object upload count
+    - Create new slice by time or count
+
+## Changes on GC
+
+- New logic in GC to support uncommitted garbage collection
```

Comment:
I just want to confirm that we acknowledge the need for a design discussion before we implement any change. I mean to answer the question of how to fit the uncommitted GC implementation into the existing GC implementation; are they two separate Spark apps? what are the implications on performance, observability, and stability of the products? how does the user runs each job, product-wise - can the jobs be separated from the user's perspective? etc.